### ISSUE TITLE
Bug: Unsupported type: <class 'asyncpg.pgproto.pgproto.UUID'> 

### ISSUE THREAD

**Issue by xgemx** [litestar-org/litestar#1785]
### Description

Can't serialize asyncpg UUID in sqlalchemy model.

ps. PR - to fix it: https://github.com/litestar-org/litestar/pull/1784

### MCVE

```python
import typing
from uuid import uuid4

import litestar
import litestar.contrib.sqlalchemy.plugins.init as sqlalchemy_plugin
import sqlalchemy
from litestar.contrib.sqlalchemy.dto import SQLAlchemyDTO
from sqlalchemy import Column, orm
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy.types import UUID


# monkey patch to fix
# litestar.serialization.DEFAULT_TYPE_ENCODERS[UUID] = str  # noqa: ERA001
SQLALCHEMY_CONFIG: typing.Final[sqlalchemy_plugin.SQLAlchemyAsyncConfig] = sqlalchemy_plugin.SQLAlchemyAsyncConfig(
    connection_string="postgresql+asyncpg://user:password@localhost/db",
)

SQLALCHEMY_PLUGIN: typing.Final[sqlalchemy_plugin.SQLAlchemyInitPlugin] = sqlalchemy_plugin.SQLAlchemyInitPlugin(
    config=SQLALCHEMY_CONFIG,
)


class Base(orm.DeclarativeBase):
    object_id: orm.Mapped[str] = Column(UUID, default=uuid4, primary_key=True, unique=True)

    @orm.declared_attr.directive
    def __tablename__(cls: "Base") -> str:  # noqa: N805, D105
        return cls.__name__.lower()


class ClientModel(Base):
    surname: orm.Mapped[str]
    firstname: orm.Mapped[str]
    patronymic: orm.Mapped[str]


ClientDTO: typing.Final[SQLAlchemyDTO] = SQLAlchemyDTO[ClientModel]


@litestar.get("/clients", dto=ClientDTO, return_dto=ClientDTO)
async def list_clients(db_session: AsyncSession) -> list[ClientModel]:
    result: typing.Final[sqlalchemy.ScalarResult] = await db_session.scalars(sqlalchemy.select(ClientModel))
    return result.all()


API_ROUTER: typing.Final[litestar.Router] = litestar.Router(
    path="/api",
    route_handlers=[list_clients],
)

litestar_application: typing.Final[litestar.Litestar] = litestar.Litestar(
    route_handlers=[API_ROUTER],
    openapi_config=litestar.openapi.OpenAPIConfig(
        title="service",
        version="1.0.0",
        description="Micro-service",
    ),
    plugins=[SQLALCHEMY_PLUGIN],
    debug=True,
)
```

### Steps to reproduce

```bash
1. Run alembic init -t async migrations 
2. Edit alembic configs to connect to db
3. Run alembic revision --autogenerate -m "init"
4. Run alembic upgrade head
5. Insert some data to the new DB (the main thing is uuid must be set)
6. Make get "api/clients" request
```

### Screenshots

```bash
"![SCREENSHOT_DESCRIPTION](SCREENSHOT_LINK.png)"
```

### Logs

```bash
Traceback (most recent call last):
  File "/usr/local/lib/python3.11/site-packages/litestar/serialization.py", line 164, in encode_json
    return _msgspec_json_encoder.encode(obj)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/litestar/serialization.py", line 123, in default_serializer
    raise TypeError(f"Unsupported type: {type(value)!r}")
TypeError: Unsupported type: <class 'asyncpg.pgproto.pgproto.UUID'>

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/usr/local/lib/python3.11/site-packages/litestar/middleware/exceptions/middleware.py", line 149, in __call__
    await self.app(scope, receive, send)
  File "/usr/local/lib/python3.11/site-packages/litestar/routes/http.py", line 77, in handle
    response = await self._get_response_for_request(
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/litestar/routes/http.py", line 129, in _get_response_for_request
    response = await self._call_handler_function(
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/litestar/routes/http.py", line 162, in _call_handler_function
    response: ASGIApp = await route_handler.to_response(app=scope["app"], data=response_data, request=request)
                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/litestar/handlers/http_handlers/base.py", line 472, in to_response
    return await response_handler(app=app, data=data, request=request, return_dto=self.resolve_return_dto())  # type: ignore
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/litestar/handlers/http_handlers/_utils.py", line 101, in handler
    return await create_response(data=data)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/litestar/handlers/http_handlers/_utils.py", line 74, in create_response
    response = response_class(
               ^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/litestar/response/base.py", line 110, in __init__
    self.body = content if isinstance(content, bytes) else self.render(content)
                                                           ^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/litestar/response/base.py", line 253, in render
    return encode_json(content, self._enc_hook)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/litestar/serialization.py", line 167, in encode_json
    raise SerializationException(str(msgspec_error)) from msgspec_error
litestar.exceptions.base_exceptions.SerializationException: Unsupported type: <class 'asyncpg.pgproto.pgproto.UUID'>
```

### Litestar Version

v2.0.0alpha7

### Platform

- [ ] Linux
- [X] Mac
- [ ] Windows
- [ ] Other (Please specify in the description above)

**Comment by jcrist** [comment]
`msgspec` author here - I _think_ the fix you've added in #1784 should solve the problem for now (though a litestar maintainer will need to confirm that :)), but I plan to push a fix upstream in `msgspec` so the patch here will eventually be unnecessary. Upstream issue filed here: https://github.com/jcrist/msgspec/issues/428.

**Comment by provinzkraut** [comment]
I'm closing this as it is definitely not a Litestar bug. 

It may be worth opening an issue with psycopg, since it seems to me that they haven't considered this edge case.
